File: api/fxhub.go
package fxhub

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"strings"
	"sync"
)

var (
	url           = "localhost:10001"
	client        = NewTcpClient(url)
	clientWriteMu sync.Mutex
	listeners     = make(map[string]func(map[string]interface{}))
	listenersMu   sync.Mutex
)

// --- TcpClient replacement ---
type TcpClient struct {
	host string
	port string
	conn net.Conn
	mu   sync.Mutex
}

func NewTcpClient(url string) *TcpClient {
	parts := strings.Split(url, ":")
	return &TcpClient{host: parts[0], port: parts[1]}
}

func (c *TcpClient) connect() error {
	if c.conn == nil {
		conn, err := net.Dial("tcp", c.host+":"+c.port)
		if err != nil {
			return err
		}
		c.conn = conn
	}
	return nil
}

func (c *TcpClient) SendAsHttp(data map[string]interface{}, path string) (string, error) {
	if err := c.connect(); err != nil {
		return "", err
	}
	body, _ := json.Marshal(data)
	req := fmt.Sprintf("POST %s HTTP/1.1\r\nHost: %s\r\nContent-Type: application/json\r\nContent-Length: %d\r\n\r\n%s",
		path, c.host, len(body), string(body))
	c.mu.Lock()
	defer c.mu.Unlock()
	_, err := c.conn.Write([]byte(req))
	if err != nil {
		return "", err
	}
	reader := bufio.NewReader(c.conn)
	var sb strings.Builder
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			break
		}
		sb.WriteString(line)
		if strings.Contains(sb.String(), "\r\n\r\n") {
			break
		}
	}
	return sb.String(), nil
}

func (c *TcpClient) ListenAsHttp(callback func(string), path string) error {
	if err := c.connect(); err != nil {
		return err
	}
	req := fmt.Sprintf("GET %s HTTP/1.1\r\nHost: %s\r\nAccept: text/event-stream\r\n\r\n", path, c.host)
	c.mu.Lock()
	_, err := c.conn.Write([]byte(req))
	c.mu.Unlock()
	if err != nil {
		return err
	}
	reader := bufio.NewScanner(c.conn)
	for reader.Scan() {
		callback(reader.Text())
	}
	return nil
}

// --- fxhub functions ---

func SendHttp(cmd string, data map[string]interface{}) map[string]interface{} {
	res := make(map[string]interface{})
	res["sended"] = false
	clientWriteMu.Lock()
	raw, err := client.SendAsHttp(data, "/"+cmd)
	clientWriteMu.Unlock()
	res["sended"] = true
	if err != nil {
		res["success"] = false
		res["error"] = err.Error()
		return res
	}
	parts := strings.SplitN(raw, "\r\n\r\n", 2)
	if len(parts) != 2 {
		parts = strings.SplitN(raw, "\n\n", 2)
	}
	if len(parts) != 2 {
		res["success"] = false
		res["error"] = "bad HTTP format"
		return res
	}
	var parsed map[string]interface{}
	if err := json.Unmarshal([]byte(parts[1]), &parsed); err != nil {
		res["success"] = false
		res["error"] = "could not parse JSON"
		return res
	}
	for k, v := range parsed {
		res[k] = v
	}
	return res
}

func SendEvent(appId, eventType string, async bool, callback func(map[string]interface{})) map[string]interface{} {
	f := func() map[string]interface{} {
		data := map[string]interface{}{"app-id": appId, "type": eventType}
		res := SendHttp("send", data)
		if callback != nil {
			callback(res)
		}
		return res
	}
	if async {
		go f()
		return nil
	}
	return f()
}

func SetState(appId string, state interface{}, async bool, callback func(map[string]interface{})) map[string]interface{} {
	f := func() map[string]interface{} {
		data := map[string]interface{}{"app-id": appId, "state": state}
		res := SendHttp("set-state", data)
		if callback != nil {
			callback(res)
		}
		return res
	}
	if async {
		go f()
		return nil
	}
	return f()
}

func GetState(appId string, async bool, callback func(map[string]interface{})) map[string]interface{} {
	f := func() map[string]interface{} {
		data := map[string]interface{}{"app-id": appId}
		res := SendHttp("state", data)
		if callback != nil {
			callback(res)
		}
		return res
	}
	if async {
		go f()
		return nil
	}
	return f()
}

func AddListener(appId, eventType string, cb func(map[string]interface{})) {
	listenersMu.Lock()
	defer listenersMu.Unlock()
	listeners[appId+"_"+eventType] = cb
}

func Listen() {
	online := func(line string) {
		if !strings.HasPrefix(line, "data:") {
			return
		}
		var payload map[string]interface{}
		if err := json.Unmarshal([]byte(line[5:]), &payload); err != nil {
			return
		}
		id := payload["app-id"].(string) + "_" + payload["type"].(string)
		listenersMu.Lock()
		cb := listeners[id]
		listenersMu.Unlock()
		if cb != nil {
			cb(payload)
		}
	}
	_ = client.ListenAsHttp(online, "/sse")
}

func ListenAsync() {
	go Listen()
}

File: api/fxhub.rs
use std::collections::HashMap;
use std::io::{Read, Write, BufRead, BufReader};
use std::net::TcpStream;
use std::sync::{Mutex, Arc};
use std::thread;
use serde_json::Value;

lazy_static::lazy_static! {
    static ref URL: String = "localhost:10001".to_string();
    static ref CLIENT: Mutex<TcpClient> = Mutex::new(TcpClient::new(&URL));
    static ref CLIENT_WRITE: Mutex<()> = Mutex::new(());
    static ref LISTENERS: Mutex<HashMap<String, Box<dyn Fn(Value) + Send + Sync>>> = Mutex::new(HashMap::new());
}

struct TcpClient {
    host: String,
    port: u16,
    conn: Option<TcpStream>,
}

impl TcpClient {
    fn new(url: &str) -> Self {
        let parts: Vec<&str> = url.split(':').collect();
        let host = parts[0].to_string();
        let port = parts[1].parse::<u16>().unwrap();
        TcpClient { host, port, conn: None }
    }

    fn connect(&mut self) {
        if self.conn.is_none() {
            self.conn = Some(TcpStream::connect((&*self.host, self.port)).unwrap());
        }
    }

    fn send_as_http(&mut self, data: &Value, path: &str) -> String {
        self.connect();
        let body = data.to_string();
        let req = format!(
            "POST {} HTTP/1.1\r\nHost: {}\r\nContent-Type: application/json\r\nContent-Length: {}\r\n\r\n{}",
            path, self.host, body.len(), body
        );
        let conn = self.conn.as_mut().unwrap();
        conn.write_all(req.as_bytes()).unwrap();
        let mut buf = String::new();
        conn.read_to_string(&mut buf).unwrap();
        buf
    }

    fn listen_as_http(&mut self, callback: Arc<dyn Fn(String) + Send + Sync>, path: &str) {
        self.connect();
        let req = format!(
            "GET {} HTTP/1.1\r\nHost: {}\r\nAccept: text/event-stream\r\n\r\n",
            path, self.host
        );
        let conn = self.conn.as_mut().unwrap();
        conn.write_all(req.as_bytes()).unwrap();

        let reader = BufReader::new(conn.try_clone().unwrap());
        for line in reader.lines() {
            if let Ok(l) = line {
                callback(l);
            }
        }
    }
}

fn send_http(cmd: &str, data: Value) -> Value {
    let mut res = serde_json::json!({"sended": false});
    let _guard = CLIENT_WRITE.lock().unwrap();
    let mut client = CLIENT.lock().unwrap();
    let raw = client.send_as_http(&data, &format!("/{}", cmd));
    res["sended"] = Value::Bool(true);

    let parts: Vec<&str> = raw.split("\r\n\r\n").collect();
    if parts.len() != 2 {
        return serde_json::json!({
            "success": false,
            "error": "Bad HTTP format"
        });
    }
    match serde_json::from_str(parts[1]) {
        Ok(parsed) => {
            let mut parsed_map: Value = parsed;
            if let Some(obj) = parsed_map.as_object_mut() {
                for (k, v) in obj.iter() {
                    res[k] = v.clone();
                }
            }
        }
        Err(_) => {
            res["success"] = Value::Bool(false);
            res["error"] = Value::String("Couldn't parse JSON".to_string());
        }
    }
    res
}

fn add_listener(app_id: &str, event_type: &str, callback: Box<dyn Fn(Value) + Send + Sync>) {
    let mut map = LISTENERS.lock().unwrap();
    map.insert(format!("{}_{}", app_id, event_type), callback);
}

fn listen() {
    let callback = Arc::new(|line: String| {
        if !line.starts_with("data:") {
            return;
        }
        if let Ok(payload) = serde_json::from_str::<Value>(&line[5..]) {
            let id = format!(
                "{}_{}",
                payload["app-id"].as_str().unwrap(),
                payload["type"].as_str().unwrap()
            );
            let map = LISTENERS.lock().unwrap();
            if let Some(cb) = map.get(&id) {
                cb(payload);
            }
        }
    });
    let mut client = CLIENT.lock().unwrap();
    client.listen_as_http(callback, "/sse");
}

fn listen_async() {
    thread::spawn(|| listen());
}
